File: container-with-most-water/container-with-most-water-mustafasegf.go
package containerwithmostwater

func maxArea(height []int) int {
	ans := 0
	l, r := 0, len(height)-1

	for l < r {
		t := min(height[l], height[r]) * (r - l)
		ans = max(ans, t)

		if height[l] <= height[r] {
			l++
		} else {
			r--
		}
	}
	return ans
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: container-with-most-water/container-with-most-water-mustafasegf.rs
use std::cmp::{max, min};

impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let mut left = 0;
        let mut right = height.len() - 1;

        let mut ans = 0;

        while left < right {
            let area = min(height[left], height[right]) * (right - left) as i32;
            ans = max(area, ans);

            match height[left] <= height[right] {
                true => left += 1,
                false => right -= 1,
            }
        }
        return ans;
    }
}

